File: core/src/vct/col/ast/ParallelRegion.java
package vct.col.ast;

import java.util.Arrays;

public class ParallelRegion extends ASTNode {
  
  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
    }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }

  public final ParallelBlock blocks[];
  public final Contract contract; 
  
  public ParallelRegion(Contract contract,ParallelBlock ... blocks){
    this.blocks=Arrays.copyOf(blocks,blocks.length);
    this.contract=contract;
  }

}

File: vercors/src/main/java/vct/col/ast/stmt/composite/ParallelRegion.scala
package vct.col.ast.stmt.composite

import scala.collection.JavaConverters._
import vct.col.ast.generic.ASTNode
import vct.col.ast.stmt.decl.Contract
import vct.col.ast.util.ASTVisitor
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}

case class ParallelRegion(val contract:Contract, val blocks:List[ParallelBlock]) extends ASTNode with VisitorHelper {
  require(blocks != null, "The list of blocks is null.")

  /** Constructs a new parallel region from an array of blocks. */
  def this(contract:Contract, blocks:Array[ParallelBlock]) = this(contract, blocks.toList)

  /** Constructs a new parallel region from Java constructs (for Java interoperability). */
  def this(contract:Contract, blocks:java.util.List[ParallelBlock]) = this(contract, blocks.asScala.toList)

  /** Provides a Java wrapper (as `java.util.List`) over the list of blocks. */
  def blocksJava = blocks.asJava

  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))

  override def debugTreeChildrenFields(): Iterable[String] = Seq("contract", "blocks")
  override def debugTreePropertyFields(): Iterable[String] = Seq()
}
